frame/uniques/src/functions.rs

```diff
@@ -195,8 +201,85 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 
 		Item::<T, I>::remove(&collection, &item);
 		Account::<T, I>::remove((&owner, &collection, &item));
+		ItemPriceOf::<T, I>::remove(&collection, &item);
 
 		Self::deposit_event(Event::Burned { collection, item, owner });
 		Ok(())
 	}
+
+	pub fn do_set_price(
+		collection: T::CollectionId,
+		item: T::ItemId,
+		sender: T::AccountId,
+		price: Option<ItemPrice<T, I>>,
+		whitelisted_buyer: Option<T::AccountId>,
+	) -> DispatchResult {
+		let collection_details =
+			Collection::<T, I>::get(&collection).ok_or(Error::<T, I>::UnknownCollection)?;
+		ensure!(!collection_details.is_frozen, Error::<T, I>::Frozen);
+		ensure!(!T::Locker::is_locked(collection, item), Error::<T, I>::Locked);
+
+		let details = Item::<T, I>::get(&collection, &item).ok_or(Error::<T, I>::UnknownItem)?;
+		ensure!(!details.is_frozen, Error::<T, I>::Frozen);
+		ensure!(details.owner == sender, Error::<T, I>::NoPermission);
+
+		if let Some(ref price) = price {
+			ItemPriceOf::<T, I>::insert(
+				&collection,
+				&item,
+				(price.clone(), whitelisted_buyer.clone()),
+			);
+			Self::deposit_event(Event::ItemPriceSet {
```

Comment:
I think we should only emit an even if the price changed.  
This would incur an additional storage read though.

Reply:
This isnt worth it imo. I believe that the difference between just writing here vs read then write will increase the PoV size and stuff like that. Chances are, user is not uselessly setting the price to be the same, and if they do, we don't care.